### DISCUSSION TITLE
Bank divider after value rather than adding more

### DISCUSSION THREAD

**Discussion by gringow831** [tttapa/Control-Surface#399] [General]
Hi Peter. I am trying to make an additional control surface working on MCU with banks where I already have a few potentiometer with two buttons (increment and decrease). 
The first bank starts on knob 1 to 8, the second goes from 9 till 16. Now, is there any way to make the 3rd Bank start at 5 to 12 rather than going back to 1 after I press to increase from 9/16?

> 1 to 8 > 9 to 16 > 5 to 12 and  increasing again brings me back to 1-8 or decrease to 9-16

Thanks

**Comment by tttapa** [comment]
Please post the code you have right now. You'll probably want to use the Many Addresses Bankable MIDI elements: https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/dc0/group__ManyAddressesMIDIOutputElements.html

**Reply by gringow831** [reply]
I am trying the many addresses as you mentioned I've tried the many addresses and when using midi I can actually do the task but over MCU there's no VOLUME_9 unless it's over a bank to control the 9th rotary fader and so on.

So far this would have been the case but only get it to work 1-8 9-16 and goes back to 1 rather thank going 5.
Bank<3> bank(8); // A bank with four channels, and 2 bank settings

IncrementDecrementSelector<3> selector = {
  bank, {2, 3},  Wrap::Wrap,
};

Bankable::PBPotentiometer knobs[] = {
  {{bank, BankType::CHANGE_CHANNEL}, A0, MCU::VOLUME_1},
  {{bank, BankType::CHANGE_CHANNEL}, A1, MCU::VOLUME_2},
  {{bank, BankType::CHANGE_CHANNEL}, A2, MCU::VOLUME_3},
  {{bank, BankType::CHANGE_CHANNEL}, A3, MCU::VOLUME_4},
  {{bank, BankType::CHANGE_CHANNEL}, 9, MCU::VOLUME_5},
  {{bank, BankType::CHANGE_CHANNEL}, 8, MCU::VOLUME_6},
  {{bank, BankType::CHANGE_CHANNEL}, 10, MCU::VOLUME_7},
  {{bank, BankType::CHANGE_CHANNEL}, 4, MCU::VOLUME_8},
};


I managed to use the manyaddresses element but I can't go to VOLUME_9 over MCU as it doesn't exist.
Bank<3> bank(8);
IncrementDecrementSelector<3> bankSelector = {bank, {2, 3}, Wrap::Wrap};

Bankable::ManyAddresses::PBPotentiometer<3> knobs[] = {
  { bank, A0, {MCU::VOLUME_1, MCU::VOLUME_9, MCU::VOLUME_5,}},
  { bank, A0, {MCU::VOLUME_2, MCU::VOLUME_10, MCU::VOLUME_6,}},
  { bank, A0, {MCU::VOLUME_3, MCU::VOLUME_11, MCU::VOLUME_7,}},
  { bank, A0, {MCU::VOLUME_4, MCU::VOLUME_12, MCU::VOLUME_8,}},
  { bank, A0, {MCU::VOLUME_5, MCU::VOLUME_13, MCU::VOLUME_9,}},
  { bank, A0, {MCU::VOLUME_6, MCU::VOLUME_14, MCU::VOLUME_10,}},
  { bank, A0, {MCU::VOLUME_7, MCU::VOLUME_15, MCU::VOLUME_11,}},
  { bank, A0, {MCU::VOLUME_8, MCU::VOLUME_16, MCU::VOLUME_12,}},
};

**Reply by tttapa** [reply]
The MCU protocol only has 8 volume controls for individual tracks, controlled by pitch bend messages on channels 1 through 8. The 9th channel controls the master volume. AFAIK, the other channels are not in use.

If you can use channel 9 as the volume for a specific track, you're probably not actually using the MCU protocol. In that case, simply use the `CHANNEL_1` through `CHANNEL_16` constants instead of `MCU::VOLUME_x`.

**Comment by gringow831** [comment]
I actually managed to do it with the following way and it does the trick

Bankable::ManyAddresses::PBPotentiometer<3> knobs[] = {
  { bank, A0, {MCU::VOLUME_1, MCU::VOLUME_1 + 8, MCU::VOLUME_1 + 4,}},
  { bank, A1, {MCU::VOLUME_2, MCU::VOLUME_2 + 8, MCU::VOLUME_2 + 4,}},
  { bank, A2, {MCU::VOLUME_3, MCU::VOLUME_3 + 8, MCU::VOLUME_3 + 4,}},
  { bank, A3, {MCU::VOLUME_4, MCU::VOLUME_4 + 8, MCU::VOLUME_4 + 4,}},
  { bank, 9, {MCU::VOLUME_5, MCU::VOLUME_5 + 8, MCU::VOLUME_5 + 4,}},
  { bank, 8, {MCU::VOLUME_6, MCU::VOLUME_6 + 8, MCU::VOLUME_6 + 4,}},
  { bank, 10, {MCU::VOLUME_7, MCU::VOLUME_7 + 8, MCU::VOLUME_7 + 4,}},
  { bank, 4, {MCU::VOLUME_8, MCU::VOLUME_8 + 8, MCU::VOLUME_8 + 4,}},
};

Is there a way over MCU to have more than 16 of potentiometers? let's say I have a second control surface this one starting from potentiometer 17 or adding it to the actual code for the same control surface? 
I tried adding +16 but MIDI has 16 channels only.

**Reply by tttapa** [reply]
Yes, this is a limitation of MIDI, there are only 16 channels, so only 16 distinct pitch bend controls.  
If you need more, either use a different type of message like control change (7-bit or 14-bit), or use multiple USB-MIDI virtual cables. This is only supported on Teensy using the MIDI x4 and and MIDI x16 USB types, and you can use the `CABLE_1` through `CABLE_16` constants in your MIDI addresses, e.g. 
```cpp
Bankable::ManyAddresses::PBPotentiometer<3> knob = {
    bank, A0, 
    {{
        {CHANNEL_1, CABLE_2},
        {CHANNEL_9, CABLE_2},
        {CHANNEL_5, CABLE_2},
    }},
};
```
Note the double pair of braces around the address list.

**Reply by gringow831** [reply]
Thank you so much.

I remember reading about it on https://github.com/tttapa/Control-Surface/issues/215#issuecomment-643650314 and https://github.com/tttapa/Control-Surface/issues/147#issuecomment-605671653 but can't put it in practice

I am still learning how to modify most of the parameters. I guess even if I weren't using the MCU protocol I couldn't control more that 16 potentiometer for pitchbend or is there any other way to control 17 to 24 without a Teensy x3 or x4?

**Reply by tttapa** [reply]
You could update the USB descriptors in the MIDI USB library for the board you're using so it supports multiple cables. A single MIDI link only has 16 pitch bend channels, AFAIK, there's no way around it.

**Reply by gringow831** [reply]
I've been trying to make use of what you mentioned but haven't been able to make it work unfortunately. I used PBPotentiomenter without the MCU protocol and they work well up to channel 8 too but nothing after 8. 
I guess if what I am trying to make is an Mackie Control Extender style controller I would need to look at using a Teensy rather than a Leonardo to change cable or am I wrong?

**Reply by tttapa** [reply]
If your DAW thinks that it's talking to a Mackie Control Universel, you can only use the first 8 channels for volume controls. If you want to be able to use all 16, you have to use different mapping in your DAW (i.e. non-MCU). If you want to use more than 8 volume controls using MCU, using extenders is the only option (apart from using the MCU bank controls of course).

If you're interested in how USB descriptors work, you could update the MIDIUSB code for the Leonardo to add some extra MIDI endpoints (not sure what the limitation here is, I don't know how many endpoints you can have on a Leonardo). You can find the USB MIDI 1.0 spec here: https://www.midi.org/specifications/midi-transports-specifications/usb  
Look in the appendices, first, there are some examples, the actual specification is a bit hard to digest.
  
If you just want something that works with multiple MIDI endpoints, I'd recommend getting a Teensy (beware, not all Teensies support multiple cables, Teensy 3.x and Teensy 4.x support it, I think the Teensy LC also, but not Teensy 2.x).

**Reply by gringow831** [reply]
Thanks Pieter!

The controller I am trying to build controls a machine rather than a DAW and it works mainly with MCU protocol. It can accept some midi signal.

The machine itself has 4 knobs to control all 16 channels so I am trying to work out a way of having extra knobs to control 5 to 10 and 11 to 16 with 2 banks.

I have tried MCU::BANK_LEFT and MCU::BANK_RIGHT but this moves the 8 channels at a time and MCU::CHANNEL_LEFT or MCU::CHANNEL_RIGHT has been disabled as it is in the machine unfortunately and it controls something completely different. Using CHANNEL_9 onwards also doesn't work but from 1 to 8 it does.

I have managed to control 1 to 8 and with a BANK_RIGHT and LEFT 9 to 16 but not 5 to 10 for example with the 6 knobs. (It does work on a DAW with IncrementDecrementSelector and ManyAddresses but not on the machine) so if I do that I am missing 5 to 8 so far so I am going around thinking what it could be done to achieve this.

I've tried many combinations and nothing seems to be working unfortunately. My last thought would be seeing if theres something like CHANNEL_LEFT, CHANNEL_RIGHT that is not MCU but over MIDI and maybe that works to change channels to start on channel5 onwards.

```cpp
Bank<2> bank(6);
/*IncrementDecrementSelector<2> bankSelector = {bank, {2, 3}, Wrap::Clamp};
*/

Bankable::ManyAddresses::PBPotentiometer<2> knobs[] = {
  { bank, A0, {CHANNEL_1, CHANNEL_1 + 4,}},
  { bank, A1, {CHANNEL_2, CHANNEL_2 + 4,}},
  { bank, A2, {CHANNEL_3, CHANNEL_3 + 4,}},
  { bank, A3, {CHANNEL_4, CHANNEL_4 + 4,}}, 
  { bank, 10, {CHANNEL_5, CHANNEL_5 + 4,}}, //CHANNEL5+4 doesn't work in the machine
  { bank, 6,  {CHANNEL_6, CHANNEL_6 + 4,}}, //CHANNEL6+4 doesn't work in the machine
};

/*
Bankable::PBPotentiometer notmanyaknob[] = {
  {{bank, BankType::CHANGE_CHANNEL}, A0, CHANNEL_1},
  {{bank, BankType::CHANGE_CHANNEL}, A1, CHANNEL_2},
  {{bank, BankType::CHANGE_CHANNEL}, A2, CHANNEL_3},
  {{bank, BankType::CHANGE_CHANNEL}, A3, CHANNEL_4},
  {{bank, BankType::CHANGE_CHANNEL}, 10, CHANNEL_5},
  {{bank, BankType::CHANGE_CHANNEL}, 6, CHANNEL_6},
};
*/

/* Bankable::ManyAddresses::PBPotentiometer<2> knobs[] = {
  { bank, A0, {VOLUME_1, VOLUME_1 + 4,}},
  { bank, A1, {VOLUME_2, VOLUME_2 + 4,}},
  { bank, A2, {VOLUME_3, VOLUME_3 + 4,}},
  { bank, A3, {VOLUME_4, VOLUME_4 + 4,}},
  { bank, 10, {VOLUME_5, VOLUME_5 + 4,}}, //VOLUME_5+4 doesn't work in the machine
  { bank, 6,  {VOLUME_6, VOLUME_6 + 4,}}, //VOLUME_6+4 doesn't work in the machine
}; 
*/

NoteButton noteButton[] = {
  { 2, MCU::BANK_LEFT,},
  { 3, MCU::BANK_RIGHT,},
};

/*Channel move doesn't work over MCU in the machine
NoteButton noteButton[] = {
  { 2, MCU::CHANNEL_LEFT,},
  { 3, MCU::CHANNEL_RIGHT,},
};
*/
```

**Reply by tttapa** [reply]
> The controller I am trying to build controls a machine rather than a DAW and it works mainly with MCU protocol. It can accept some midi signal.

Then surely this mystery machine must have a manual with the MIDI implementation and a list of all messages it supports?

> I have tried MCU::BANK_LEFT and MCU::BANK_RIGHT but this moves the 8 channels at a time

Why is this an issue? Simply keep track of which MCU bank you've selected and then before sending a MIDI pitch bend message check whether you have to change the bank first using BANK_LEFT/RIGHT.

> My last thought would be seeing if theres something like CHANNEL_LEFT, CHANNEL_RIGHT that is not MCU but over MIDI and maybe that works to change channels to start on channel5 onwards.

Not that I know of, you'll have to consult the manual of your device to be sure, but I doubt it.
